### PR TITLE
Feature/fct monthly active students

### PR DESCRIPTION
**Pull request by jordan-springer**
# Description

Update, test, and validate existing [`fct_active_students_monthly`](https://github.com/code-dot-org/analytics/blob/main/dbt/models/marts/students/dim_active_students.sql) with proposed model design and updated upstream model-feed.

image

Model runtime: ~ 7 min

Summary statistics: 
 - average ~22% change in num of active students (having removed anonymous users from the former model)
image

## Links

Jira ticket(s): [DATAOPS-956](https://codedotorg.atlassian.net/browse/DATAOPS-956)

## Testing story

- [x] Does your change include appropriate tests on key columns?
      eg.
      - `not_null`
      - `unique`
      - `dbt_utils.unique_combination_of_columns: , ["value","value","value"...]

**Note: when submitting a new model for review please make sure the following have been tested:**

1. The model compiles (`dbt build -m 'your_model'`)
         or: _has the dbt Cloud job succeeded?_
3. The model runs (`dbt run -m 'your_model'`)
4. The model produces accessible data in the DW (`select 1 from 'your_model'`)

## Privacy

- [ ] 1.	Does this change involve the collection, use, or sharing of new Personal Data?
- [x] 2.    Do these data exist in the appropriate schema(s)? 
- [ ] 3.	Does this change involve a new or changed use or sharing of existing Personal Data?
- [x] 4.    Consider: will this data be visible on Tableau? will this data be surfaced in a report exported from Trevor?
- [ ] 5.    If yes to any of the above, please list the models, columns, and justification below:
      i.
      ii.
      iii. 

## PR Checklist:
--> **Note: if these are not all checked, the PR will be sent back.**

- [x] Tests provide adequate coverage
- [x] Privacy and Security impacts have been assessed
- [x] Code adheres to [style guide](https://docs.getdbt.com/best-practices/how-we-style/0-how-we-style-our-dbt-projects)👀 and is **[DRY](https://docs.getdbt.com/terms/dry)**
- [x] Code is well-commented (**please do not leave extraneous commentary in model code, if it is for the purpose of documentation, please relocate accordingly)
- [x] Appropriate documentation has been provided (see `.yml.`, did `dbt docs generate` succeed?)
- [x] New features are translatable or updates will not break up/downstream models
- [x] Relevant documentation has been added or updated (i.e. `dbt docs` has been updated successfully on [Github Pages](code-dot-org.github.io/analytics/)
- [x] Pull Request is labeled appropriately (eg. `chore/`, `feature/`, `fix/`)
- [x] Follow-up work items (including potential tech debt) are tracked and linked (if applicable)

### CODE REVIEW

**Review by allison-code-dot-org — CHANGES_REQUESTED** [review]
Overall, I'm on-board with the logic behind defining an "active student", and should be relatively easy to integrate the anonymous/ signed-out usage once we have it. I left one main comment on considerations of the model's grain level and some other smaller notes to consider. Thank you for all your work on this!

File: dbt/macros/get_cutoff_date.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
+{% macro get_cutoff_date() %}
+
+    -- Define the cutoff date dynamically
+    '2019-06-30'  -- You can replace this with a dynamic value, e.g., a variable or config
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/macros/get_cutoff_date.sql:5]: I'm this used as the standard for our `/marts` offering going forward

[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/macros/get_cutoff_date.sql:5]: re-writing to be last of of school_year - 5 would be more optimal for sure, we should do that

File: dbt/models/marts/students/dim_student_projects.sql [hunk 2]
```diff
@@ -40,57 +54,57 @@ international_contact_info as (
...
-left join international_partners_raw                                        as ipr 
-    on pc.display_name = ipr.display_name
+),
+
+final as (
+    select distinct 
+          p.project_id 
+        , u.user_id
+        , case when p.user_id   is not null then 1 else 0 end   as known_cdo_user -- (1=student, 0=anon)
+        , case when u.user_id   is not null then 1 else 0 end   as is_signed_in
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/students/dim_student_projects.sql:64]: yes -- there are 3 rows, all without a user_id
```
select *
from dbt_jordan.dim_student_projects
where is_signed_in <> known_cdo_user
```

File: dbt/models/marts/metrics/fct_active_students_monthly.sql [hunk 3]
```diff
@@ -1,16 +1,134 @@
...
+        on uni.activity_date 
+            between sy.started_at 
+                and sy.ended_at
+
+    {{ dbt_utils.group_by(5) }} 
+), 
+
+final as (
+    select
+        activity_month,
```
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/models/marts/metrics/fct_active_students_monthly.sql:102]: i agree, i am also going to make a yearly model of the same for that purpose.

[the remark this answers is not in the document]
Comment (coryamanda) [reply dbt/models/marts/metrics/fct_active_students_monthly.sql:92]: I think Allison makes a good point here. The most common granularity is  for monthly roll-ups, but there are going to be times when we need to adjust that - I'm 100% certain that we will get asked for the same breakouts by week, for different countries, for different age ranges, etc. The problem with dropping the user/day model is that we have no flexibility to meet those asks. For instance, the 2030 goals include breakouts for URG and gender, as well as for curriculum target area. While this approach avoids a large dim_active_students model, we'd end up having to create separate fct aggregates for every breakout or subset of this data we might need. Having a table at the active user/day model that we can aggregate as needed meets a much broader set of needs.

[the remark this answers is not in the document]
Comment (coryamanda) [reply dbt/models/marts/metrics/fct_active_students_monthly.sql:92]: @jordan-springer see here

**Comment by jordan-springer** [comment]
runtime for new `dim_active_students` ~44 min

image

File: dbt/macros/get_cutoff_date.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (coryamanda) [reply dbt/macros/get_cutoff_date.sql:5]: cool - does it make sense to use this for the cutoff date in dim_active_students as well, or is this only for marts?

**Comment by jordan-springer** [comment]
checks passed; updated model timing 
image

File: dbt/macros/get_cutoff_date.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (jordan-springer) [reply dbt/macros/get_cutoff_date.sql:5]: that's a good question. we can start with dim_active_students and the metrics/ mart and decide what else could be covered?

**Review by jordan-springer — COMMENTED** [review]
looks good, thanks! we can now also remove this logic from the dim_active_students model

**Review by jordan-springer — COMMENTED** [review]
👍🏽

File: dbt/models/marts/students/dim_active_students.sql [hunk 4]
```diff
@@ -1,92 +1,106 @@
...
+            between sy.started_at
+                and sy.ended_at 
+
+    {{ dbt_utils.group_by(5) }} 
+),
+
+final as (
+    select *, 1 as is_active_student
+    from combined 
+    where has_sign_in_activity = 1
```
Comment (coryamanda) [inline dbt/models/marts/students/dim_active_students.sql:100]: I think this aggregation is off. When we say 'students must be signed in', that means they can't be anonymous students in the data. This filter is checking to see whether they also have a sign in event on that day. That's not the same thing - a sign in event is an authentication, but we know that many students stay signed in on their browsers for days at a time and that's fine. I think the right filter is cdo_known_user, but you should check that.

**Review by coryamanda — CHANGES_REQUESTED** [review]

File: dbt/models/marts/students/dim_active_students.sql [hunk 5]
```diff
@@ -1,92 +1,106 @@
...
+        user_id,
+        created_date    as activity_date,
+        1               as has_user_level_activity,
+        null            as has_sign_in_activity,
+        null            as has_project_activity
+        
+    from {{ ref('dim_user_levels') }}
+    where 
+        total_attempts > 0 
+        and created_date > {{ get_cutoff_date() }}
```
Comment (coryamanda) [inline dbt/models/marts/students/dim_active_students.sql:23]: filter out HOC only usage here. I asked Natalia for her recommended approach here: https://codedotorg.slack.com/archives/C071T8M847J/p1727702582301659

Comment (coryamanda) [reply dbt/models/marts/students/dim_active_students.sql:23]: select distinct course_name
from dev.analytics.dim_course_structure dcs 
where dcs.is_active_student_course = 1

File: dbt/models/marts/students/dim_active_students.sql [hunk 6]
```diff
@@ -1,92 +1,106 @@
...
+        project_created_at::date    as activity_date,
+        null                        as has_user_level_activity,
+        null                        as has_sign_in_activity,
+        1                           as has_project_activity
+
+    from {{ ref('dim_student_projects') }}
+    where 
+        user_type = 'student'
+        and project_created_at > {{ get_cutoff_date() }} 
+
```
Comment (coryamanda) [inline dbt/models/marts/students/dim_active_students.sql:55]: Add "where user_id is not null" filter here to remove the anonymous projects. Or use the "known_cdo_user" field you created in dim_student_projects.

**Review by coryamanda — CHANGES_REQUESTED** [review]

**Comment by coryamanda** [comment]
@jordan-springer A few more updates:
- We need the urg and gender fields added to dim_active_students as we discussed this morning. Just documenting that.
- Remove the filter on course for the user levels join. We can include HOC activity for the overall active students metric

File: dbt/models/marts/students/dim_active_students.sql (hunk 5 above)
Comment (jordan-springer) [reply dbt/models/marts/students/dim_active_students.sql:23]: removing this

**Comment by jordan-springer** [comment]
we are also removing sign_in's from active metrics consideration and adding back the `dim_user_course_activity model` .sorry for all the file changes!

**Review by coryamanda — APPROVED** [review]

**Comment by coryamanda** [comment]
Merging this in without waiting for Allison to re-review, given that her feedback was addressed.

**Pull request merged**